### PR TITLE
throttling down mechanism

### PR DESCRIPTION
**Pull request by bradymiller**
throttling down mechanism

### CODE REVIEW

File: src/Common/Session/SessionTracker.php [hunk 1]
```diff
@@ -64,4 +64,40 @@ public static function updateSessionExpiration(): void
...
+    // Function to throttle down requests when using the online demos to prevent abuse of the demo farm
+    public static function processSessionThrottleDown(): void
+    {
+        $throttleDownWaitMilliseconds = getenv('THROTTLE_DOWN_WAIT_MILLISECONDS', true) ?? 0;
+        if ($throttleDownWaitMilliseconds > 0 ) {
+            // calculate $timeThrottle['time_throttle'], which will be average time (in milliseconds) per script call
+            $timeThrottle = sqlQueryNoLog("SELECT `number_scripts`, `created`, NOW() as `current_timestamp` FROM `session_tracker` WHERE `uuid` = ?", $_SESSION['session_database_uuid']);
+            $timeThrottle['time_throttle'] = ((new \DateTime($timeThrottle['created']))->format('Uv') + ((int)$throttleDownWaitMilliseconds * $timeThrottle['number_scripts'])) - (new \DateTime($timeThrottle['current_timestamp']))->format('Uv');
+
+            error_log("DEBUG: timeThrottle is " . $timeThrottle['time_throttle'] . " milliseconds and number scripts is " . $timeThrottle['number_scripts']);
```
Comment (bradymiller) [inline src/Common/Session/SessionTracker.php:86]: will remove above line when bring this in

File: docker/development-easy/docker-compose.yml [hunk 2]
```diff
@@ -31,6 +31,8 @@ services:
     - logvolume:/var/log
     - couchdbvolume:/couchdb/data
     environment:
+      THROTTLE_DOWN_WAIT_MILLISECONDS: 500
```
Comment (bradymiller) [inline docker/development-easy/docker-compose.yml:34]: think 800 prob best here

File: src/Common/Session/SessionTracker.php [hunk 3]
```diff
@@ -64,4 +64,40 @@ public static function updateSessionExpiration(): void
...
+        }
+    }
+
+    // Function to throttle down requests when using the online demos to prevent abuse of the demo farm
+    public static function processSessionThrottleDown(): void
+    {
+        $throttleDownWaitMilliseconds = getenv('THROTTLE_DOWN_WAIT_MILLISECONDS', true) ?? 0;
+        if ($throttleDownWaitMilliseconds > 0 ) {
+            // calculate $timeThrottle['time_throttle'], which will be average time (in milliseconds) per script call
+            $timeThrottle = sqlQueryNoLog("SELECT `number_scripts`, `created`, NOW() as `current_timestamp` FROM `session_tracker` WHERE `uuid` = ?", $_SESSION['session_database_uuid']);
```
Comment (stephenwaite) [inline src/Common/Session/SessionTracker.php:83]: made this feat look much easier than ip blocking :sunglasses: 

doesn't the bind have to be in an array for the new adodb thing that popped up the other day?

File: src/Common/Session/SessionTracker.php [hunk 4]
```diff
@@ -64,4 +64,35 @@ public static function updateSessionExpiration(): void
...
+    }
+
+    // Function to throttle down requests when using the online demos to prevent abuse of the demo farm
+    public static function processSessionThrottleDown($throttleDownWaitMilliseconds): void
+    {
+        // calculate $timeThrottle['time_throttle'], which will be average time (in milliseconds) per script call
+        $timeThrottle = sqlQueryNoLog("SELECT `number_scripts`, `created`, NOW() as `current_timestamp` FROM `session_tracker` WHERE `uuid` = ?", $_SESSION['session_database_uuid']);
+        $timeThrottle['time_throttle'] = ((new \DateTime($timeThrottle['created']))->format('Uv') + ((int)$throttleDownWaitMilliseconds * $timeThrottle['number_scripts'])) - (new \DateTime($timeThrottle['current_timestamp']))->format('Uv');
+
+        error_log("DEBUG: timeThrottle is " . $timeThrottle['time_throttle'] . " milliseconds and number scripts is " . $timeThrottle['number_scripts']);
```
Comment (bradymiller) [inline src/Common/Session/SessionTracker.php:82]: will remove above line before bringing into codebase

File: src/Common/Session/SessionTracker.php (hunk 3 above)
Comment (bradymiller) [reply src/Common/Session/SessionTracker.php]: goodness, i copied that code from another block. Odd that it is working. Will fix it here and in original code.

Comment (bradymiller) [reply src/Common/Session/SessionTracker.php]: willing to bet that adodb is a bit flexible on this when 1 parameter, however bet breaks if it is null or empty

File: docker/development-easy/docker-compose.yml [hunk 5]
```diff
@@ -31,6 +31,8 @@ services:
...
+      THROTTLE_DOWN_DIE_MILLISECONDS: 10000
```
[the remark this answers is not in the document]
Comment (bradymiller) [reply docker/development-easy/docker-compose.yml:35]: Gonna go with 700

**Comment by bradymiller** [comment]
testing well and going to incorporate into the demo farm. hopefully this helps.

**Comment by stephenwaite** [comment]
bird_jive

File: sql/7_0_2-to-7_0_3_upgrade.sql [hunk 6]
```diff
@@ -120,4 +120,8 @@ INSERT INTO `supported_external_dataloads` (`load_type`, `load_source`, `load_re
 
 #IfMissingColumn form_encounter ordering_provider_id
 ALTER TABLE `form_encounter` ADD `ordering_provider_id` INT(11) DEFAULT '0' COMMENT 'ordering provider, if any, for this visit';
-#EndIf
\ No newline at end of file
+#EndIf
+
+#IfMissingColumn session_tracker `number_scripts`
```
Comment (bradymiller) [inline sql/7_0_2-to-7_0_3_upgrade.sql:125]: sigh, need to remove the tics above :)

**Pull request merged**